File: app/src/main/java/com/czbix/v2ex/ui/loader/TopicLoader.java
package com.czbix.v2ex.ui.loader;

import android.content.Context;

import com.czbix.v2ex.common.exception.ConnectionException;
import com.czbix.v2ex.common.exception.RemoteException;
import com.czbix.v2ex.dao.TopicDao;
import com.czbix.v2ex.model.Comment;
import com.czbix.v2ex.model.Topic;
import com.czbix.v2ex.model.TopicWithComments;
import com.czbix.v2ex.network.RequestHelper;
import com.google.common.collect.Lists;

import java.util.Collections;
import java.util.List;

public class TopicLoader extends AsyncTaskLoader<TopicWithComments> {
    private final Topic mTopic;
    private int mPage;

    public TopicLoader(Context context, Topic topic) {
        super(context);

        mTopic = topic;
        mPage = 1;
    }

    public void setPage(int page) {
        if (page == mPage) {
            return;
        }

        mPage = page;
        mResult = null;
        onContentChanged();
    }

    @Override
    public TopicWithComments loadInBackgroundWithException() throws ConnectionException, RemoteException {
        final TopicWithComments topicWithComments = RequestHelper.getTopicWithComments(mTopic, mPage);
        if (mPage == 1 && topicWithComments.mComments.size() > 0) {
            final int lastRead = TopicDao.getLastReadReply(mTopic.getId());
            if (lastRead > 7) {
                final List<Integer> commentIds = Lists.transform(topicWithComments.mComments,
                        Comment::getFloor);
                final int index = Collections.binarySearch(commentIds, lastRead);
                // the closest and latest comment of last read pos
                topicWithComments.mLastReadPos = index >= 0 ? index : -index - 2;
            }
        }

        return topicWithComments;
    }
}

File: app/src/main/kotlin/com/czbix/v2ex/ui/loader/TopicLoader.kt
package com.czbix.v2ex.ui.loader

import android.content.Context
import com.czbix.v2ex.common.exception.ConnectionException
import com.czbix.v2ex.common.exception.RemoteException
import com.czbix.v2ex.dao.TopicDao
import com.czbix.v2ex.model.Topic
import com.czbix.v2ex.model.TopicWithComments
import com.czbix.v2ex.network.RequestHelper
import java.util.*

class TopicLoader(context: Context, private val mTopic: Topic) : AsyncTaskLoader<TopicWithComments>(context) {
    private var mPage: Int = 0

    init {
        mPage = 1
    }

    fun setPage(page: Int) {
        if (page == mPage) {
            return
        }

        mPage = page
        mResult = null
        onContentChanged()
    }

    @Throws(ConnectionException::class, RemoteException::class)
    override fun loadInBackgroundWithException(): TopicWithComments {
        val topicWithComments = RequestHelper.getTopicWithComments(mTopic, mPage)
        if (mPage == 1 && topicWithComments.mComments.size > 0) {
            val lastRead = TopicDao.getLastReadReply(mTopic.id)
            if (lastRead > 7) {
                val commentIds = topicWithComments.mComments.map { it.floor }
                val index = Collections.binarySearch(commentIds, lastRead)
                // the closest and latest comment of last read pos
                topicWithComments.mLastReadPos = if (index >= 0) index else -index - 2
            }
        }

        return topicWithComments
    }
}
